### ISSUE TITLE
Search for a concurrent map with weak keys

### ISSUE THREAD

**Issue by felixbarny** [elastic/apm-agent-java#357]
Requirements
- Weak keys
- Thread safe (Not `WeakHashMap`)
- Highly performant in multi-threaded scenarios (not Guava or `Collections.synchronizedMap`)
- No allocations on `get()` (not [`WeakConcurrentMap`](https://github.com/raphw/weak-lock-free/blob/master/src/main/java/com/blogspot/mydailyjava/weaklockfree/WeakConcurrentMap.java#L51))

**Comment by felixbarny** [comment]
Relates to #317

**Comment by tylerbenson** [comment]
Good luck.  You basically excluded all the options I am aware of.

**Comment by felixbarny** [comment]
Hehe, thx :smile:
What I want to test is the map backing caffeine cache.

**Comment by ben-manes** [comment]
Caffeine is like `WeakConcurrentMap` by allocating a wrapper for the key comparison. I think this isn't a problem because of escape analysis, since the object is on the same thread. But I don't recall testing that and escape analysis may not work across method cals. However the allocation should be cheap due to short lived. Your best bet is to write a JMH benchmark. The big difference is that Caffeine does not require create any dedicated threads and performs maintenance cycles after a threshold of operations occur.

Guava gets around that by forking the hash table so that it can use a pluggable `Equivalence`. However, that is a big commitment which they did in a misguided enthusiasm towards soft references. In parallel I was exploring concurrent LRU and, when reality set in on the Guava team, we tried to port my work into theirs. So Guava is optimized for reference caching first, whereas Caffeine is size first. We could fork the hash table just for reference caching, but I really don't want to commit towards maintaining even more complexity.

**Comment by felixbarny** [comment]
Thanks a lot for your input.

> I think this isn't a problem because of escape analysis, since the object is on the same thread.

In my benchmarks, the escape analysis didn't seem to kick in, unfortunately.

> However the allocation should be cheap due to short lived.

True, but if an agent increases the allocation rate of short-lived objects, it may push the GC over its limits when the monitored application itself allocates so much that the GC is just able to keep up. In such situations, short-lived objects might end up in the old gen, increasing the frequency and duration of major GCs. That's one reason why I'm striving for a garbage free agent.

I just did some multi-threaded benchmarks for recording a servlet request and a JDBC call. The `WeakConcurrentMap` had a significantly better 99,9%'ile but increased the allocation rate by 32 bytes (from around 1 byte). Other percentiles had a too high run-to-run variance to be significant. The 99.9%'ile also had a high variance but it is always significantly better, by a factor of around 5-10 (around 100µs vs 1000µs). The 99%'ile maybe was also a bit better but not too significant.

In summary, it would be nice to have the advantages of the two approaches - a garbage-free weak concurrent map which performs well under high contention.

I guess when going the route of having a pluggable `Equivalence`, the implementation can't be based on `ConcurrentHashMap` but would have to be completely custom, right?

**Comment by felixbarny** [comment]
The results of one particular run: http://jmh.morethan.io/?sources=https://gist.githubusercontent.com/felixbarny/34dbb685a63ca56c55ec3d2dc5177151/raw/4e57b0c59505c5bd84410213af90039adbfd6350/synchronized-map.json,https://gist.githubusercontent.com/felixbarny/9149e0f368794702228c812459f0b514/raw/5adf9729bb1e1665ad199f803c76bd434c77e5f8/weak-concurrent.json

**Comment by felixbarny** [comment]
Hey, that's a really good idea! Seems to work well.

> That’s not ideal in most cases

Could you explain why this may not be ideal?

**Comment by ben-manes** [comment]
Actually I was just reconsidering that as blind bias and maybe we should do this in Caffeine too. ThreadLocals shouldn’t be abused, but yeah it does seem okay here.
